HMS-9066: Updating go version to 1.24

mayurilahane opened the pull request.

mayurilahane committed:
chore: update Go version, lint config, and improve HTTP response handling

- Bump Go version to 1.24 and golangci-lint to v2.4.0 in GitHub Actions workflow
- Update .golangci.yaml for compatibility with golangci-lint v2
- Refactor internal/zestwrapper/rpm.go to use safer HTTP response body closing with error handling and improved logging
- Switch to math/rand/v2 in pkg/tangy/queries.go for randomness, remove golang.org/x/exp/rand (deprecated) dependency
- Update dependancies in go.mod and go.sum

rverdile reviewed (commented):
I would do a dependency bump here: `go get -u ./...`

mayurilahane commented:
Getting lint errors which suggest to use the pattern to set verison is 1.24.x 

and i have set it to 1.24 which seems a better option to me

rverdile commented:
It's a little confusing, but the linter errors are actually caused by this line here: https://github.com/content-services/tang/blob/main/.github/workflows/tang-actions.yaml#L23

It needs to be updated to a newer version of golangci-lint. I would say v2.4 to match the backend: https://github.com/content-services/content-sources-backend/blob/main/.github/workflows/content-sources-actions.yml#L51-L53

mayurilahane commented:
> It's a little confusing, but the linter errors are actually caused by this line here: https://github.com/content-services/tang/blob/main/.github/workflows/tang-actions.yaml#L23
> 
> It needs to be updated to a newer version of golangci-lint. I would say v2.4 to match the backend: https://github.com/content-services/content-sources-backend/blob/main/.github/workflows/content-sources-actions.yml#L51-L53

how do you figure it out ? that it should match with backend golangci-lint version ?
i thought it will update by itself after `go get -u`  cmd

rverdile commented:
> how do you figure it out ? that it should match with backend golangci-lint version ?
i thought it will update by itself after go get -u cmd

Good question. `go get -u ./...` will only upgrade the dependencies imported in the go.mod. golangci-lint is not a dependency, but a CLI tool for linting. We install and use it locally with `make lint`. so everything probably worked locally. However, the golangci-lint action is a separate github action, independent of what we use locally. We should maybe make some change to the action to match what we use locally, but just bumping the version is good for now :)

rverdile commented:
there's a CI error. sounds like you just need to bump the action version
```
Failed to run: Error: invalid version string 'v2.4.0', golangci-lint v2 is not supported by golangci-lint-action v6, you must update to golangci-lint-action v7., Error: invalid version string 'v2.4.0', golangci-lint v2 is not supported by golangci-lint-action v6, you must update to golangci-lint-action v7
```

mayurilahane commented:
/retest

marusak commented:
> - Bump Go version to 1.24 and golangci-lint to v2.4.0 in GitHub Actions workflow
> - Update .golangci.yaml for compatibility with golangci-lint v2
> - Refactor internal/zestwrapper/rpm.go to use safer HTTP response body closing with error handling and improved logging
> - Switch to math/rand/v2 in pkg/tangy/queries.go for randomness, remove golang.org/x/exp/rand (deprecated) dependency
> - Update dependancies in go.mod and go.sum

There seem to be ~4 different changes? But they are all in the same commit :thinking: Are they strictly connected thus being in one commit?

rverdile commented on internal/zestwrapper/rpm.go:
```diff
@@ -99,7 +107,11 @@ func (r *RpmZest) CreateRepository(domain, name, url string) (repoHref string, r
+		if closeErr := httpResp.Body.Close(); closeErr != nil {
+			fmt.Printf("failed to close response body: %v\n", closeErr)
+		}
+	}()
```
We prefer to ignore this linter error. Can you update the PR to use the config [from the backend](https://github.com/content-services/content-sources-backend/blob/main/.golangci.yaml)? That will make sure all the right linters are turned on

rverdile commented:
> There seem to be ~4 different changes? But they are all in the same commit 🤔 Are they strictly connected thus being in one commit?

@marusak They're all connected to bumping the go version and failures related to that

mayurilahane commented:
> > * Bump Go version to 1.24 and golangci-lint to v2.4.0 in GitHub Actions workflow
> > * Update .golangci.yaml for compatibility with golangci-lint v2
> > * Refactor internal/zestwrapper/rpm.go to use safer HTTP response body closing with error handling and improved logging
> > * Switch to math/rand/v2 in pkg/tangy/queries.go for randomness, remove golang.org/x/exp/rand (deprecated) dependency
> > * Update dependancies in go.mod and go.sum
> 
> There seem to be ~4 different changes? But they are all in the same commit 🤔 Are they strictly connected thus being in one commit?

Hey yes, these are connected.
I updated go and go linter version.

Updating linter version, recomended these changes which looked good to me in pkg/tangy/queries.go and  internal/zestwrapper/rpm.go.

Although i need to undo the changes in  internal/zestwrapper/rpm.go as per 
https://github.com/content-services/tang/pull/21#discussion_r2391394979

mayurilahane committed:
fix: update linter config and resolve G115 and ST1005 linter errors

Update .golangci.yaml to match backend linter configuration to ensure
all appropriate linters are enabled.

- G115 - Integer overflow in pool limit:
  Change PoolLimit type from int to int32 to match pgxpool.Config.MaxConns,
  eliminating type conversion and preventing potential integer overflow.

- ST1005 - Error strings should not be capitalized
  Fix error message capitalization in zestwrapper/rpm.go

rverdile commented on pkg/tangy/interface.go:
```diff
@@ -19,7 +19,7 @@ func New(dbConfig Database, logConfig Logger) (Tangy, error) {
 		dbConfig.PoolLimit = DefaultMaxPoolLimit
 	}
-	pxConfig.MaxConns = int32(dbConfig.PoolLimit)
+	pxConfig.MaxConns = dbConfig.PoolLimit
```
This works! but changing types here will also require changes in the backend. Let's fix this linting issue by checking the int max, instead of changing the type of `dbConfig.PoolLimit` to int32, like this: https://github.com/content-services/content-sources-backend/blob/93863b503e71c6bb93e10f1b2239887efc476c86/pkg/tasks/queue/pgqueue.go#L218-L220

mayurilahane committed:
add: validation for pool limit bounds in New function

Validate that dbConfig.PoolLimit is within 32-bit integer range
(math.MinInt32 to math.MaxInt32) before converting to int32 to
prevent overflow issues.

rverdile reviewed (approved):
looks good! going to merge and release a new version. Then you'll be able to update your backend PR
